client/client.go

```diff
@@ -173,30 +173,99 @@ func unmarshalRepoPackages(p, cacheDir string, cacheLife time.Duration, proxySer
 	}
 	logger.Infof("Fetching repo content for %s, cache either doesn't exist or is older than %v", p, cacheLife)
 
-	url := p + "/index.gz"
-	logger.Infof("Fetching %q", url)
-	res, err := httpClient.Get(url)
+	parsedP, err := url.Parse(p)
 	if err != nil {
 		return nil, err
 	}
+	switch parsedP.Scheme {
+	case "http", "https":
+		return unmarshalRepoPackagesHTTP(parsedP, cf, proxyServer)
+	case "gs":
+		return unmarshalRepoPackagesGCS(parsedP, cf, proxyServer)
+	}
 
-	if res.StatusCode == 200 {
-		return decode(res, cf)
+	return nil, fmt.Errorf("Unsupported URL Scheme '%s' for '%s'", parsedP.Scheme, p)
+}
+
+func unmarshalRepoPackagesHTTP(repoURL *url.URL, cf string, proxyServer string) ([]goolib.RepoSpec, error) {
+	httpClient := &http.Client{}
+	if proxyServer != "" {
+		proxyURL, err := url.Parse(proxyServer)
+		if err != nil {
+			logger.Fatal(err)
+		}
+		httpClient.Transport = &http.Transport{Proxy: http.ProxyURL(proxyURL)}
 	}
 
-	logger.Infof("Gzipped index returned status: %q, trying plain JSON.", res.Status)
-	url = p + "/index"
-	logger.Infof("Fetching %q", url)
-	res, err = httpClient.Get(url)
+	indexURL := repoURL.String() + "/index.gz"
+	logger.Infof("Fetching %q", indexURL)
+	res, err := httpClient.Get(indexURL)
 	if err != nil {
 		return nil, err
 	}
 
 	if res.StatusCode != 200 {
-		return nil, fmt.Errorf("index GET request returned status: %q", res.Status)
+		logger.Infof("Gzipped index returned status: %q, trying plain JSON.", res.Status)
+		indexURL = repoURL.String() + "/index"
+		logger.Infof("Fetching %q", indexURL)
+		res, err = httpClient.Get(indexURL)
+		if err != nil {
+			return nil, err
+		}
+
+		if res.StatusCode != 200 {
+			return nil, fmt.Errorf("index GET request returned status: %q", res.Status)
+		}
+	}
+
+	index, err := ioutil.ReadAll(res.Body)
+	return decode(index, cf)
+}
+
+func unmarshalRepoPackagesGCS(repoURL *url.URL, cf string, proxyServer string) ([]goolib.RepoSpec, error) {
+
+	if proxyServer != "" {
+		logger.Errorf("Proxy server not supported with gs:// URLs, skiping repo '%s'", repoURL.String())
+		var empty []goolib.RepoSpec
+		return empty, nil
+	}
+
+	ctx := context.Background()
```

Comment:
ctx should be passed in, context.Background() should only be called in main or tests.

The Execute function in googet_download.go already has a context to use, you'll just need to pipe it down to here (also ctx should always be the first argument)

Reply:
I'll also take a TODO here, i can go through and clean this up when I run through testing

Reply:
I missed the TODO(adjakura) comment, so the latest commit plumbs the context into the new code. Sorry for the merge conflict if you had something coded but not yet pushed.